### PR TITLE
Parse `new Worker(new URL("...", import.meta.url))`

### PR DESCRIPTION
**Pull request by mischnic**
Part of https://github.com/parcel-bundler/parcel/issues/5430

`new Worker("...")` should be deprecated at some point

Bildschirmfoto 2021-02-08 um 23 40 24

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 3.13s | -132.00ms |
| Cached | 718.00ms | -0.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 711.00ms | -39.00ms 🚀 |
| dist/legacy/index.33d06174.js | 1.11kb | +0.00b | 680.00ms | +35.00ms ⚠️ |
| dist/modern/index.9e1f1458.js | 1.11kb | +0.00b | 720.00ms | -43.00ms 🚀 |
| dist/legacy/index.html | 701.00b | +0.00b | 679.00ms | +34.00ms ⚠️ |
| dist/modern/index.html | 701.00b | +0.00b | 718.00ms | -39.00ms 🚀 |
| dist/modern/index.0bc656bc.css | 77.00b | +0.00b | 1.18s | -69.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/index.33d06174.js | 1.11kb | +0.00b | 80.00ms | +19.00ms ⚠️ |
| dist/modern/index.9e1f1458.js | 1.11kb | +0.00b | 62.00ms | -26.00ms 🚀 |
| dist/legacy/index.html | 701.00b | +0.00b | 81.00ms | +19.00ms ⚠️ |
| dist/legacy/index.44edca46.css | 77.00b | +0.00b | 81.00ms | +20.00ms ⚠️ |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 23.42s | +1.24s ⚠️ |
| Cached | 9.48s | -102.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/PermalinkedComment.77f121c4.js | 4.20kb | +0.00b | 794.00ms | -263.00ms 🚀 |
| dist/UserProfile.125b5ebe.js | 1.76kb | +0.00b | 794.00ms | -264.00ms 🚀 |
| dist/NotFound.d769112d.js | 530.00b | +0.00b | 261.00ms | +16.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 483.35kb | +0.00b | 87.00ms | +21.00ms ⚠️ |
| dist/PermalinkedComment.77f121c4.js | 4.20kb | +0.00b | 87.00ms | +21.00ms ⚠️ |
| dist/UserProfile.125b5ebe.js | 1.76kb | +0.00b | 87.00ms | +21.00ms ⚠️ |
| dist/NotFound.d769112d.js | 530.00b | +0.00b | 47.00ms | +5.00ms ⚠️ |

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 2.55m | -5.90s |
| Cached | 3.40s | -107.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.d51fa13e.js | 2.46mb | +0.00b | 1.26m | -20.56s 🚀 |
| dist/pdfRenderer.ded75111.js | 1.11mb | +0.00b | 46.60s | -2.56s 🚀 |
| dist/editorView.2ca486fc.js | 587.53kb | +0.00b | 1.53m | +53.72s ⚠️ |
| dist/EmojiPickerComponent.90b741d3.js | 139.01kb | +0.00b | 13.61s | +2.07s ⚠️ |
| dist/media-viewer.d1cb7a32.js | 72.87kb | +0.00b | 12.63s | -1.07s 🚀 |
| dist/card.ecae978d.js | 54.02kb | +0.00b | 9.53s | -2.45s 🚀 |
| dist/Modal.a2c499cb.js | 39.18kb | +0.00b | 5.16s | +756.00ms ⚠️ |
| dist/component.fa4e42d1.js | 30.82kb | +0.00b | 2.19s | -1.95s 🚀 |
| dist/esm.6bc5d42c.js | 27.67kb | +0.00b | 13.66s | +2.32s ⚠️ |
| dist/component.63b8a376.js | 22.54kb | +0.00b | 5.16s | +777.00ms ⚠️ |
| dist/DatePicker.cbcf2d18.js | 21.20kb | +0.00b | 6.61s | -4.63s 🚀 |
| dist/js.4b5d27df.js | 16.55kb | +0.00b | 5.14s | +764.00ms ⚠️ |
| dist/dropzone.a58778e7.js | 16.06kb | +0.00b | 18.13s | +3.33s ⚠️ |
| dist/ui.bda48e80.js | 14.21kb | +0.00b | 9.22s | -2.40s 🚀 |
| dist/workerHasher.09e24457.js | 11.90kb | +0.00b | 13.63s | +6.30s ⚠️ |
| dist/component.f521dd8b.js | 6.20kb | +0.00b | 3.78s | -569.00ms 🚀 |
| dist/card.1a9729d8.js | 5.71kb | +0.00b | 12.31s | -947.00ms 🚀 |
| dist/media-viewer.941270ab.js | 4.01kb | +0.00b | 12.85s | -892.00ms 🚀 |
| dist/EmojiPickerComponent.ca69a3c9.js | 3.62kb | +0.00b | 9.20s | -2.16s 🚀 |
| dist/png-chunks-extract.e6351d92.js | 3.55kb | +0.00b | 5.14s | +764.00ms ⚠️ |
| dist/index.ece3a13e.css | 3.46kb | +0.00b | 21.57s | +1.91s ⚠️ |
| dist/Modal.00beb515.js | 3.14kb | +0.00b | 3.24s | +2.12s ⚠️ |
| dist/16.cf92dee6.js | 2.49kb | +0.00b | 2.17s | +1.06s ⚠️ |
| dist/ResourcedEmojiComponent.0df891ba.js | 2.15kb | +0.00b | 9.20s | -2.41s 🚀 |
| dist/date.f35453a2.js | 1.96kb | +0.00b | 5.50s | +395.00ms ⚠️ |
| dist/images.861d8ede.js | 1.90kb | +0.00b | 5.69s | +574.00ms ⚠️ |
| dist/feedback.ddd45d32.js | 1.86kb | +0.00b | 6.61s | -4.62s 🚀 |
| dist/16.4dbc75f9.js | 1.86kb | +0.00b | 5.56s | +649.00ms ⚠️ |
| dist/16.934d2c3d.js | 1.79kb | +0.00b | 4.00s | -668.00ms 🚀 |
| dist/workerHasher.afb52a72.js | 1.75kb | +0.00b | 13.61s | +6.28s ⚠️ |
| dist/list-number.95becf64.js | 1.68kb | +0.00b | 5.69s | +412.00ms ⚠️ |
| dist/status.5568e78d.js | 1.68kb | +0.00b | 5.83s | +405.00ms ⚠️ |
| dist/link.b4c4afb6.js | 1.53kb | +0.00b | 5.69s | +413.00ms ⚠️ |
| dist/heading6.b56f3cf2.js | 1.53kb | +0.00b | 6.61s | +1.02s ⚠️ |
| dist/heading3.0c0fe1f4.js | 1.51kb | +0.00b | 5.98s | +400.00ms ⚠️ |
| dist/16.d564a7d9.js | 1.46kb | +0.00b | 4.00s | -626.00ms 🚀 |
| dist/emoji.4f12a585.js | 1.45kb | +0.00b | 5.52s | +412.00ms ⚠️ |
| dist/16.81d1aa0c.js | 1.45kb | +0.00b | 2.17s | +1.04s ⚠️ |
| dist/16.897e8807.js | 1.44kb | +0.00b | 5.34s | +428.00ms ⚠️ |
| dist/16.c2dce044.js | 1.40kb | +0.00b | 5.34s | +429.00ms ⚠️ |
| dist/heading5.bde11017.js | 1.40kb | +0.00b | 6.61s | +1.02s ⚠️ |
| dist/expand.ca6be77b.js | 1.38kb | +0.00b | 6.61s | -4.62s 🚀 |
| dist/16.ac96fdd0.js | 1.36kb | +0.00b | 5.56s | +649.00ms ⚠️ |
| dist/16.620d8c15.js | 1.34kb | +0.00b | 2.17s | +1.06s ⚠️ |
| dist/heading2.1511a751.js | 1.33kb | +0.00b | 5.98s | +397.00ms ⚠️ |
| dist/16.e0980eea.js | 1.32kb | +0.00b | 6.03s | +889.00ms ⚠️ |
| dist/16.990be6f2.js | 1.31kb | +0.00b | 5.35s | +431.00ms ⚠️ |
| dist/mention.e2501ad0.js | 1.29kb | +0.00b | 6.58s | +1.29s ⚠️ |
| dist/heading4.1abfeb0a.js | 1.29kb | +0.00b | 5.98s | +398.00ms ⚠️ |
| dist/Modal.b44f65e1.js | 1.28kb | +0.00b | 5.14s | +765.00ms ⚠️ |
| dist/16.30ebc170.js | 1.27kb | +0.00b | 2.17s | +1.04s ⚠️ |
| dist/layout.6a9de36b.js | 1.27kb | +0.00b | 5.69s | +412.00ms ⚠️ |
| dist/16.3ba7b0ec.js | 1.27kb | +0.00b | 5.56s | +908.00ms ⚠️ |
| dist/16.2ec03e4f.js | 1.26kb | +0.00b | 4.33s | -561.00ms 🚀 |
| dist/16.98030c60.js | 1.26kb | +0.00b | 5.13s | +500.00ms ⚠️ |
| dist/16.c20ef86c.js | 1.26kb | +0.00b | 4.00s | -626.00ms 🚀 |
| dist/divider.9a733dfe.js | 1.25kb | +0.00b | 5.52s | +412.00ms ⚠️ |
| dist/quote.be70ca39.js | 1.25kb | +0.00b | 5.83s | +405.00ms ⚠️ |
| dist/component.3a931beb.js | 1.23kb | +0.00b | 3.24s | -861.00ms 🚀 |
| dist/16.7b2e1497.js | 1.23kb | +0.00b | 3.24s | -861.00ms 🚀 |
| dist/media-card-analytics-error-boundary.c2d3f430.js | 1.23kb | +0.00b | 12.31s | -940.00ms 🚀 |
| dist/decision.0740ec10.js | 1.21kb | +0.00b | 5.50s | +395.00ms ⚠️ |
| dist/panel-warning.9b047bba.js | 1.21kb | +0.00b | 5.83s | +408.00ms ⚠️ |
| dist/list.e4113e60.js | 1.18kb | +0.00b | 5.69s | +412.00ms ⚠️ |
| dist/heading1.cf012b35.js | 1.18kb | +0.00b | 5.98s | -1.06s 🚀 |
| dist/panel-error.0c73ee4f.js | 1.11kb | +0.00b | 6.58s | +809.00ms ⚠️ |
| dist/panel.53ea0317.js | 1.10kb | +0.00b | 5.83s | +408.00ms ⚠️ |
| dist/table.23e2141c.js | 1.09kb | +0.00b | 5.97s | +545.00ms ⚠️ |
| dist/panel-note.d5e6a337.js | 1.05kb | +0.00b | 6.58s | +809.00ms ⚠️ |
| dist/media-card-analytics-error-boundary.64c19024.js | 999.00b | +0.00b | 12.31s | -940.00ms 🚀 |
| dist/simpleHasher.931d48e8.js | 755.00b | +0.00b | 9.20s | +1.83s ⚠️ |
| dist/index.html | 119.00b | +0.00b | 2.30s | -908.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.9b7c69b7.js | 2.46mb | +4.00b ⚠️ | 690.00ms | -14.00ms |
| dist/pdfRenderer.ded75111.js | 1.11mb | +0.00b | 714.00ms | -65.00ms 🚀 |
| dist/editorView.2ca486fc.js | 587.53kb | +0.00b | 750.00ms | -38.00ms 🚀 |
| dist/popup.74263bce.js | 169.24kb | +0.00b | 728.00ms | -52.00ms 🚀 |
| dist/EmojiPickerComponent.90b741d3.js | 139.01kb | +0.00b | 701.00ms | -60.00ms 🚀 |
| dist/Toolbar.57f2e74b.js | 99.69kb | +0.00b | 750.00ms | -38.00ms 🚀 |
| dist/media-viewer.d1cb7a32.js | 72.87kb | +0.00b | 714.00ms | -64.00ms 🚀 |
| dist/card.86f6e816.js | 51.77kb | +0.00b | 738.00ms | -50.00ms 🚀 |
| dist/esm.6bc5d42c.js | 27.67kb | +0.00b | 701.00ms | -61.00ms 🚀 |
| dist/component.63b8a376.js | 22.54kb | +0.00b | 627.00ms | -67.00ms 🚀 |
| dist/DatePicker.cbcf2d18.js | 21.20kb | +0.00b | 690.00ms | -55.00ms 🚀 |
| dist/smartMediaEditor.70ca3661.js | 16.85kb | +0.00b | 750.00ms | -38.00ms 🚀 |
| dist/js.4b5d27df.js | 16.55kb | +0.00b | 627.00ms | -67.00ms 🚀 |
| dist/dropzone.0f14607d.js | 16.06kb | +0.00b | 738.00ms | -50.00ms 🚀 |
| dist/workerHasher.09e24457.js | 11.90kb | +0.00b | 701.00ms | -61.00ms 🚀 |
| dist/component.f521dd8b.js | 6.20kb | +0.00b | 627.00ms | -67.00ms 🚀 |
| dist/card.1a9729d8.js | 5.71kb | +0.00b | 714.00ms | -64.00ms 🚀 |
| dist/media-viewer.941270ab.js | 4.01kb | +0.00b | 714.00ms | -65.00ms 🚀 |
| dist/png-chunks-extract.e6351d92.js | 3.55kb | +0.00b | 627.00ms | -67.00ms 🚀 |
| dist/index.ece3a13e.css | 3.46kb | +0.00b | 753.00ms | -38.00ms 🚀 |
| dist/dropzone.ceefa141.js | 3.44kb | +0.00b | 728.00ms | -52.00ms 🚀 |
| dist/Modal.00beb515.js | 3.14kb | +0.00b | 616.00ms | -88.00ms 🚀 |
| dist/clipboard.1be7d5c2.js | 2.97kb | +0.00b | 738.00ms | -50.00ms 🚀 |
| dist/16.cf92dee6.js | 2.49kb | +0.00b | 616.00ms | -65.00ms 🚀 |
| dist/card.b6509036.js | 2.15kb | +0.00b | 728.00ms | -52.00ms 🚀 |
| dist/date.f35453a2.js | 1.96kb | +0.00b | 659.00ms | -58.00ms 🚀 |
| dist/images.861d8ede.js | 1.90kb | +0.00b | 665.00ms | -55.00ms 🚀 |
| dist/feedback.ddd45d32.js | 1.86kb | +0.00b | 690.00ms | -55.00ms 🚀 |
| dist/16.4dbc75f9.js | 1.86kb | +0.00b | 647.00ms | -65.00ms 🚀 |
| dist/browser.ba398ea7.js | 1.82kb | +0.00b | 738.00ms | -50.00ms 🚀 |
| dist/16.934d2c3d.js | 1.79kb | +0.00b | 636.00ms | -67.00ms 🚀 |
| dist/workerHasher.afb52a72.js | 1.75kb | +0.00b | 701.00ms | -61.00ms 🚀 |
| dist/workerHasher.42458950.js | 1.75kb | +0.00b | 750.00ms | -38.00ms 🚀 |
| dist/list-number.95becf64.js | 1.68kb | +0.00b | 670.00ms | -65.00ms 🚀 |
| dist/status.5568e78d.js | 1.68kb | +0.00b | 678.00ms | -61.00ms 🚀 |
| dist/code.94d6ed50.js | 1.61kb | +0.00b | 659.00ms | -58.00ms 🚀 |
| dist/link.b4c4afb6.js | 1.53kb | +0.00b | 665.00ms | -56.00ms 🚀 |
| dist/heading6.b56f3cf2.js | 1.53kb | +0.00b | 690.00ms | -54.00ms 🚀 |
| dist/heading3.0c0fe1f4.js | 1.51kb | +0.00b | 683.00ms | -70.00ms 🚀 |
| dist/16.2a10e9af.js | 1.51kb | +0.00b | 646.00ms | -65.00ms 🚀 |
| dist/16.d564a7d9.js | 1.46kb | +0.00b | 636.00ms | -67.00ms 🚀 |
| dist/16.01c01dd6.js | 1.46kb | +0.00b | 637.00ms | -67.00ms 🚀 |
| dist/emoji.4f12a585.js | 1.45kb | +0.00b | 665.00ms | -55.00ms 🚀 |
| dist/16.81d1aa0c.js | 1.45kb | +0.00b | 616.00ms | -65.00ms 🚀 |
| dist/16.897e8807.js | 1.44kb | +0.00b | 654.00ms | -58.00ms 🚀 |
| dist/16.c2dce044.js | 1.40kb | +0.00b | 654.00ms | -58.00ms 🚀 |
| dist/heading5.bde11017.js | 1.40kb | +0.00b | 689.00ms | -55.00ms 🚀 |
| dist/expand.ca6be77b.js | 1.38kb | +0.00b | 690.00ms | -55.00ms 🚀 |
| dist/16.ac96fdd0.js | 1.36kb | +0.00b | 654.00ms | -58.00ms 🚀 |
| dist/16.620d8c15.js | 1.34kb | +0.00b | 616.00ms | -65.00ms 🚀 |
| dist/heading2.1511a751.js | 1.33kb | +0.00b | 683.00ms | -70.00ms 🚀 |
| dist/16.e0980eea.js | 1.32kb | +0.00b | 654.00ms | -58.00ms 🚀 |
| dist/16.990be6f2.js | 1.31kb | +0.00b | 659.00ms | -58.00ms 🚀 |
| dist/mention.e2501ad0.js | 1.29kb | +0.00b | 671.00ms | -64.00ms 🚀 |
| dist/heading4.1abfeb0a.js | 1.29kb | +0.00b | 683.00ms | -70.00ms 🚀 |
| dist/Modal.b44f65e1.js | 1.28kb | +0.00b | 626.00ms | -68.00ms 🚀 |
| dist/16.30ebc170.js | 1.27kb | +0.00b | 616.00ms | -65.00ms 🚀 |
| dist/layout.6a9de36b.js | 1.27kb | +0.00b | 665.00ms | -55.00ms 🚀 |
| dist/16.3ba7b0ec.js | 1.27kb | +0.00b | 647.00ms | -65.00ms 🚀 |
| dist/16.2ec03e4f.js | 1.26kb | +0.00b | 637.00ms | -67.00ms 🚀 |
| dist/16.b9cdc131.js | 1.26kb | +0.00b | 646.00ms | -65.00ms 🚀 |
| dist/16.98030c60.js | 1.26kb | +0.00b | 647.00ms | -64.00ms 🚀 |
| dist/16.c20ef86c.js | 1.26kb | +0.00b | 636.00ms | -67.00ms 🚀 |
| dist/divider.9a733dfe.js | 1.25kb | +0.00b | 665.00ms | -55.00ms 🚀 |
| dist/quote.be70ca39.js | 1.25kb | +0.00b | 678.00ms | -59.00ms 🚀 |
| dist/action.88cd1e67.js | 1.23kb | +0.00b | 659.00ms | -58.00ms 🚀 |
| dist/media-card-analytics-error-boundary.c2d3f430.js | 1.23kb | +0.00b | 714.00ms | -64.00ms 🚀 |
| dist/decision.0740ec10.js | 1.21kb | +0.00b | 659.00ms | -58.00ms 🚀 |
| dist/panel-warning.9b047bba.js | 1.21kb | +0.00b | 678.00ms | -59.00ms 🚀 |
| dist/16.c890a474.js | 1.18kb | +0.00b | 654.00ms | -58.00ms 🚀 |
| dist/list.e4113e60.js | 1.18kb | +0.00b | 670.00ms | -65.00ms 🚀 |
| dist/heading1.cf012b35.js | 1.18kb | +0.00b | 683.00ms | -70.00ms 🚀 |
| dist/panel-error.0c73ee4f.js | 1.11kb | +0.00b | 671.00ms | -64.00ms 🚀 |
| dist/panel.53ea0317.js | 1.10kb | +0.00b | 678.00ms | -59.00ms 🚀 |
| dist/table.23e2141c.js | 1.09kb | +0.00b | 682.00ms | -70.00ms 🚀 |
| dist/panel-success.b56c18fb.js | 1.05kb | +0.00b | 678.00ms | -59.00ms 🚀 |
| dist/panel-note.d5e6a337.js | 1.05kb | +0.00b | 671.00ms | -64.00ms 🚀 |
| dist/media-picker-analytics-error-boundary.16608041.js | 1003.00b | +0.00b | 728.00ms | -52.00ms 🚀 |
| dist/media-card-analytics-error-boundary.60a98f4f.js | 999.00b | +0.00b | 738.00ms | -50.00ms 🚀 |
| dist/media-viewer-analytics-error-boundary.57caec83.js | 995.00b | +0.00b | 728.00ms | -52.00ms 🚀 |
| dist/simpleHasher.931d48e8.js | 755.00b | +0.00b | 701.00ms | -61.00ms 🚀 |
| dist/simpleHasher.d4526885.js | 755.00b | +0.00b | 750.00ms | -38.00ms 🚀 |

</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 16.28s | +251.00ms |
| Cached | 736.00ms | -6.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/289917092278305289)**

File: packages/transformers/js/src/visitors/dependencies.js [hunk 1]
```diff
@@ -8,24 +8,51 @@ import type {
...
+  isReturnStatement,
+  isStringLiteral,
 } from '@babel/types';
 import {isURL, md5FromString, createDependencyLocation} from '@parcel/utils';
 import {isInFalsyBranch, hasBinding, morph} from './utils';
 
 const serviceWorkerPattern = ['navigator', 'serviceWorker', 'register'];
 
+function parseImportMetaUrl(node: Node): ?string {
+  if (isNewExpression(node) && isIdentifier(node.callee, {name: 'URL'})) {
```
Comment (devongovett) [inline packages/transformers/js/src/visitors/dependencies.js:40]: Check that URL is not declared in a local scope?

File: packages/transformers/js/src/visitors/dependencies.js [hunk 2]
```diff
@@ -188,8 +216,43 @@ export default ({
           meta: {
             webworker: true,
           },
-        });
-        return;
+        };
+
+        if (isStringLiteral(args[0])) {
+          addURLDependency(asset, ast, args[0], opts);
```
Comment (devongovett) [inline packages/transformers/js/src/visitors/dependencies.js:222]: Maybe we could add a warning here? IMO we should probably remove it before stable parcel 2, so a deprecation warning in the meantime would be good.

File: packages/transformers/js/src/visitors/dependencies.js [hunk 3]
```diff
@@ -188,8 +216,43 @@ export default ({
...
+          return;
+        } else {
+          let url = parseImportMetaUrl(args[0]);
+          if (url) {
+            let {loc} = args[0];
+            if (loc) {
+              opts = {
+                ...opts,
+                loc: {
```
Comment (devongovett) [inline packages/transformers/js/src/visitors/dependencies.js:231]: Why not convertBabelLoc?

**Comment by devongovett** [comment]
Should we also do this for service worker?

File: packages/transformers/js/src/visitors/dependencies.js (hunk 2 above)
Comment (mischnic) [reply packages/transformers/js/src/visitors/dependencies.js:222]: Bildschirmfoto 2021-02-11 um 19 30 43

**Review by wbinnssmith — APPROVED** [review]

**Review by devongovett — APPROVED** [review]

**Pull request merged**